### ISSUE TITLE
Feature request dedicated occ scan job

### ISSUE THREAD

**Issue by derritter88** [nextcloud/recognize#496]
As I still have some issues with recognize + cron.php I would like to have the possiblitiy to disable/enable recognize from regular Nextcloud cronjob.

My intention would be to initially run a recognize scan for everything which takes longer.
After the initial recognize job has finished all new images could then be done via the regular cron.php

**Comment by phil-lipp** [comment]
Same here. Some kind of feature to facilitate the initial scan would be very much appreciated!

**Comment by marcelklehr** [comment]
There is an open request for this at #432 

I recognize the need (no pun intended), but it's

**Comment by derritter88** [comment]
> There is an open request for this at #432
> 
> I recognize the need (no pun intended), but it's rather complex to support both modi of operation, and in that case I believe the better modus is the cron job. Which problems do you have with the cron job specifically? Maybe I can fix them.

My problem is/was that today my nextcloud server CPU had 100 % as it was full of cron.php jobs which didn't do anything.
I think that probably the classifiers takes more time to identify x photos (x = value which classfier jobs will process all 5 minutes).
In general I run my cron.php all 5 minutes.

So if cron.php from 5 minutes ago is still running due to classifying going on another one will be started and will also do classifying which leads to a deadlock.

My Nextcloud handles around 500k images of all users.


That's why I personal would like to have a possbility to not use recognize within cron.php but as a standalone occ possibility which I can then use from regular crontab.

**Comment by phil-lipp** [comment]
hi @derritter88 
I have / had the same problem. A bit of searching online brought up a similar topic with the bookmarks app. A user recommended the command `sudo killall -q -u www-data --older-than 5m php` to be run in intervals. It did the trick for me for the last couple of days. 
@marcelklehr in my opinion the initial scan of large amount of pictures with lots of faces is the main reason for having an occ command. I have close to 100.000 photos and running it little by little (like 50 pictures day) would take literally ages.
Isn't it possible to evoke the same command that the cron.php does for classification just with a huge batch size? Or something similar?

**Comment by derritter88** [comment]
@phil-lipp 
hello philipp, sure that might be a solution (kill cron.php) but personal I am not a big fan of killing a process where I do not know why it hangs or for which task completion it waits.

I have disabled recognize for the moment as it also kills my server even with a very limited set of data per 5 minutes.

**Comment by marcelklehr** [comment]
Which batch size have you configured in the admin settings?

**Comment by derritter88** [comment]
@marcelklehr Initially I left it at the default values which caused the issues with cpu/cron.php
 Later as I wanted to get the job done I disabled them and played around with higher values and manually starting the cron.php this worked a bit better.

But I also saw that if there was nothing more to recognize although there were alot of entries at the Db and from the batch size it should also have continued to do its job it just stopped doing anything.
cron.php used 1 CPU core at max and I ctrl+c it and disabled the app.

**Comment by marcelklehr** [comment]
if you have performance problems, setting the batch size to lower values should help

**Comment by derritter88** [comment]
No it's not a perfomance issue as far as I can see - it just creates one cron.php job after another. From my point of view one cron.php waits for something to be finished, the next one waits for the previous one to be finished and so on.

But I do not know on which process the 1st cron.php waits

**Comment by marcelklehr** [comment]
@derritter88 Do you want to hop onto https://gitter.im/marcelklehr/recognize to chat a bit. Might be faster :)

**Comment by derritter88** [comment]
Okay so after playing around with settings/values my personal best practice is to lower the values to 50 % of the default values.
After that it works pretty flawless without any issues like a hanging cron.php job.

**Comment by bugsyb** [comment]
@derritter88 , @marcelklehr 

Not a fix, though as some of my cron jobs take more than 5 mins to finish (Recognize), I came up with below - allowing to run in fact total of two cron jobs, as realized that time to load it from the storage/conversion and then tensorflow get best use of all resources if two jobs are run in parallel.

This is quick & dirty hackish approach, though works perfectly.
I have cron calling this cron-wrapper.sh every minute for initial full scan of +~100k photos. With this approach I have always one job for sure running tensorflow and other is initiating pull, etc. In my case I have this offloaded to stronger system. NC DB (postgres) and storage is sitting on small SBC (N2) and all Recognize processing I have currently done in parallel from two stronger systems which gives me a real kick and I can connect more easily. It just is missing GPU which I'd be keen to use too.

```
#!/bin/bash
LOG=/tmp/.cron.log
STILLRUNNING=$(ps aux|grep "/bin/docker exec -t -u www-data nextcloud /bin/bash"|grep -v grep|wc -l)
if [ $STILLRUNNING -ge 2 ]; then
  echo "$$ - $(date) - still running other 2 other processes - exiting" >> $LOG
  exit 1
fi
echo "#######" >> $LOG
echo "$$ - $(date) started" >> $LOG
start=`date +%s`
starts=`date`
# check if running and if yes, exit
/bin/docker exec -t -u www-data nextcloud /bin/bash -c "/usr/bin/nice -n 19 php -f /var/www/html/cron.php"
end=`date +%s`
echo "$$ - $(date) finished in $((end-start)) seconds (started $starts)" >> $LOG

```

**Comment by phil-lipp** [comment]
> @derritter88 , @marcelklehr
> 
> Not a fix, though as some of my cron jobs take more than 5 mins to finish (Recognize), I came up with below - allowing to run in fact total of two cron jobs, as realized that time to load it from the storage/conversion and then tensorflow get best use of all resources if two jobs are run in parallel. This avoids killing process ([#496 (comment)](https://github.com/nextcloud/recognize/issues/496#issuecomment-1315657883)) as awaits for it to be nicely finished.
> 
> This is quick & dirty hackish approach, though works perfectly. I have cron calling this cron-wrapper.sh every minute for initial full scan of +~100k photos. With this approach I have always one job for sure running tensorflow and other is initiating pull, etc. In my case I have this offloaded to stronger system. NC DB (postgres) and storage is sitting on small SBC (N2) and all Recognize processing I have currently done in parallel from two stronger systems which gives me a real kick and I can connect more easily. It just is missing GPU which I'd be keen to use too.
> 
> ```
> #!/bin/bash
> LOG=/tmp/.cron.log
> STILLRUNNING=$(ps aux|grep "/bin/docker exec -t -u www-data nextcloud /bin/bash"|grep -v grep|wc -l)
> if [ $STILLRUNNING -ge 2 ]; then
>   echo "$$ - $(date) - still running other 2 other processes - exiting" >> $LOG
>   exit 1
> fi
> echo "#######" >> $LOG
> echo "$$ - $(date) started" >> $LOG
> start=`date +%s`
> starts=`date`
> # check if running and if yes, exit
> /bin/docker exec -t -u www-data nextcloud /bin/bash -c "/usr/bin/nice -n 19 php -f /var/www/html/cron.php"
> end=`date +%s`
> echo "$$ - $(date) finished in $((end-start)) seconds (started $starts)" >> $LOG
> ```

Hey @bugsyb I'd like to try out your approach, especially the part of offloading the heavy load to a beafier machine than that one that's running the Nextcloud (a Raspberry Pi in my case). How did you set up Recognize on the stronger system and what's your way of syncing the db (from stronger machine to SBC)?

**Comment by bugsyb** [comment]
@phil-lipp , sure, happy to share the logic.

The key is to have:

1. postgres/mysql based database for NC, as concurrent sessions are required and Sqlite3 isn't probably very happy with this. Take precautions, as that I didn't check source code if there are semaphores to allow multiple sessions, though everything works for quite some time for me.
2. Storage folder exposed via, i.e. nfs, to allow multiple systems accessing same storage pool (user files).

With above fulfilled, on beefier machine, it's all down to mounting storage and given database server is the same, it just works ;)

Well, there's small quirk. If you'd want to have that beefier machine serving all over the web, then all apps need to be installed again, as per DB entries, NC will believe apps are installed, but will be missing on the file-system. There's also one more quirk, I've noticed, if apps are getting into de-sync mode, one flipping between systems, NC will ask to perform upgrade, which is pretty annoying (takes a minute or so, as it runs DB checks, etc.).

For simple Recognize run, it's literally down to installing NC and adding, i.e. from CLI Recognize app and cron job to be run.
Then it just works, at least for me. Process of tagging all data hasn't finished for me yet though is pretty promising.

Why there shouldn't be any issues with concurrent tagging, etc.?
There are no mechanisms from NC side blocking multiple cron jobs (for good or bad) and one could end up with multiple threads from same system too, effectively running multiple sessions to DB and file-system.

Hope that makes sense.

In my case, I'm running it out of docker and... well, normally system is on small SBC, which is good normally. Recently I did come up to the point that given often I run pretty powerful system most of the days, at that time I wanted to have the more powerful one taking all requests, as this is the time one family also uses the NC server. Therefore there's Traefik load-balancer, in fact on that small SBC, which if discovers that beefy system is up, it uses it as front-end, effectively providing also HA solution. When the powerful one goes down, it gets discovered and UI, etc. is all served by the small boy.

Having that setup and effectively having all ingredients when discovering that Recognize might take years to process my family NC (~2.3M just image files), would take probably weeks if not months on small boy, beefier system has been plugged. This was in hopes to get GPU utilized, though upon discovery Recognize doesn't use GPU yet.
With this done and realizing it still could be quicker, I now have set of 3 beefy systems working on the files and tagging it all in parallel. 
The best I found is running up to two processes per HW system to use best of resources/time as it takes time to load data prior NN/Tensorflow jumps in, therefore when one abuses CPU, the other preps data for processing, hence running it in distance of a minute (cron run - above example).

If any questions would arise, let me know, I'll be happy to share setup details.
p.s. on default NC container, there's no "nice" command - I've NC container modded with small additions.
